multiuser/api/che-multiuser-api-authentication-commons/src/main/java/org/eclipse/che/multiuser/api/authentication/commons/DestroySessionListener.java

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright (c) 2012-2018 Red Hat, Inc.
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.multiuser.api.authentication.commons;
+
+import static org.eclipse.che.multiuser.api.authentication.commons.Constants.CHE_SUBJECT_ATTRIBUTE;
+
+import com.google.inject.Injector;
+import javax.servlet.ServletContext;
+import javax.servlet.http.HttpSessionEvent;
+import javax.servlet.http.HttpSessionListener;
+import org.eclipse.che.commons.subject.Subject;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** Purges deleted sessions from sessions cache store. */
+public class DestroySessionListener implements HttpSessionListener {
+
+  private static final Logger LOG = LoggerFactory.getLogger(DestroySessionListener.class);
+
+  @Override
+  public final void sessionCreated(HttpSessionEvent se) {}
+
+  @Override
+  public void sessionDestroyed(HttpSessionEvent se) {
+
+    ServletContext servletContext = se.getSession().getServletContext();
+
+    SessionStore sessionStore = getInstance(SessionStore.class, servletContext);
+    if (sessionStore == null) {
+      LOG.error(
+          "Unable to remove session from store. Session store is not configured in servlet context.");
+      return;
+    }
+
+    Subject subject = (Subject) se.getSession().getAttribute(CHE_SUBJECT_ATTRIBUTE);
+    if (subject != null) {
+      sessionStore.remove(subject.getUserId());
+    }
+  }
+
+  /** Searches component in servlet context when with help of guice injector. */
+  private <T> T getInstance(Class<T> type, ServletContext servletContext) {
```

Comment:
Does it have to use generics if it is private method called only once with `SessionStore` class?

Reply:
+0 more general code looks better and can be copy-pasted easily :rofl:

Reply:
I disagree, it's one hit private method. And you don't want to copy-paste it. Also if you use `SessionStore.class` directly, you get rid of unchecked cast.

Reply:
+1 for @sparkoo